kad-dht/README.md

```diff
@@ -0,0 +1,288 @@
+# libp2p Kademlia DHT specification
+
+The Kademlia Distributed Hash Table (DHT) subsystem in libp2p is a DHT
+implementation largely based on the Kademlia [0] whitepaper, augmented with
+notions from S/Kademlia [1], Coral [2] and mainlineDHT \[3\].
+
+This specification assumes the reader has prior knowledge of those systems. So
+rather than explaining DHT mechanics from scratch, we focus on differential
+areas:
+
+1. Specialisations and peculiarities of the libp2p implementation.
+2. Actual wire messages.
+3. Other algorithmic or non-standard behaviours worth pointing out.
+
+For everything else that isn't explicitly stated herein, it is safe to assume
+behaviour similar to Kademlia-based libraries.
+
+Code snippets use a Go-like syntax.
+
+## Authors
+
+* Protocol Labs.
+
+## Editors
+
+* [Raúl Kripalani](https://github.com/raulk)
+* [John Hiesey](https://github.com/jhiesey)
+
+## Distance function (dXOR)
+
+The libp2p Kad DHT uses the **XOR distance metric** as defined in the original
+Kademlia paper [0]. Peer IDs are normalised through the SHA256 hash function.
+
+For recap, `dXOR(sha256(id1), sha256(id2))` is the number of common leftmost
+bits between SHA256 of each peer IDs. The `dXOR` between us and a peer X
+designates the bucket index that peer X will take up in the Kademlia routing
+table.
+
+## Kademlia routing table
+
+The data structure backing this system is a k-bucket routing table, closely
+following the design outlined in the Kademlia paper [0]. The default value for
+`k` is 20, and the maximum bucket count matches the size of the SHA256 function,
+i.e. 256 buckets.
+
+The routing table is unfolded lazily, starting with a single bucket a position 0
+(representing the most distant peers), and splitting it subsequently as closer
+peers are found, and the capacity of the nearmost bucket is exceeded.
+
+## Alpha concurrency factor (α)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
```

Comment:
I'm not sure why the spec has this. In real-world implementations, a concurrency factor much, much higher is required to be reasonably fast.

Reply:
You're right, this probably doesn't belong in the spec. However, with the implementation of https://github.com/libp2p/go-libp2p-kad-dht/issues/146 landing, the number of outgoing requests is multiplied by a factor of the number of paths (currently default of 10). That will change the behavior in practice.

We'll have to do some testing to determine what this should be set to.

@raulk what do you think?

Reply:
For now, I would like to keep the concurrency factor default value at `3`. After all, implementations may diverge from the default values suggested in this specification. I think it is well worth it to extend this specification with a more in-depth discussion on query concurrency, though I would prefer this discussion to happen in a follow-up pull request.

Please let me know in case you object.